types.go

```diff
@@ -15,33 +15,39 @@ import (
 
 const defaultGradientAbsTol = 1e-6
 
-// EvaluationType is used by the optimizer to specify information needed
-// from the objective function.
-type EvaluationType int
+// EvaluationType is used by the optimizer to specify information needed from
+// the objective function. The types can be composed together using the binary
+// or operator.
+type EvaluationType uint
 
+// Basic evaluation types.
 const (
-	NoEvaluation EvaluationType = iota
-	FuncEvaluation
+	NoEvaluation   EvaluationType = 0
+	FuncEvaluation EvaluationType = 1 << iota
 	GradEvaluation
 	HessEvaluation
-	FuncGradEvaluation
-	FuncGradHessEvaluation
+)
+
+// Combined evaluation types.
+const (
+	FuncGradEvaluation     = FuncEvaluation | GradEvaluation
+	FuncGradHessEvaluation = FuncGradEvaluation | HessEvaluation
 )
 
 func (e EvaluationType) String() string {
-	if e < 0 || int(e) >= len(evaluationStrings) {
-		return fmt.Sprintf("EvaluationType(%d)", e)
+	if s, ok := evaluationStrings[e]; ok {
+		return s
 	}
-	return evaluationStrings[e]
+	return fmt.Sprintf("EvaluationType(%d)", e)
 }
 
-var evaluationStrings = [...]string{
-	"NoEvaluation",
-	"FuncEvaluation",
-	"GradEvaluation",
-	"HessEvaluation",
-	"FuncGradEvaluation",
-	"FuncGradHessEvaluation",
+var evaluationStrings = map[EvaluationType]string{
```

Comment:
Would you say that "EvaluationType(7)" is more meaningful? It's likely we'll only have one more EvaluationType, which is NonlinearConstraints (other constraints are fixed, and at the moment I'm imagining that linear and non-linear are part of one function call). This is 16 combinations which is enumerable if we wanted to keep evaluationStrings. We could also return something like
````
EnumerationType(V: 11, F: true, G: true, H: false, C: true, E: 0)
EnumerationType(V: 43, F: true, G: true, H: false, C: true, E: 32)
````
where "E" is extra and would help users who get an "unknown evaluation type" panic.